src/core_api.cpp

```diff
@@ -1693,3 +1695,244 @@ bool del_preset(const std::shared_ptr<http_req>& req, const std::shared_ptr<http
+
+    if ((req_json.count("ip_addresses") == 0 && req_json.count("api_keys") == 0) || (req_json.count("ip_addresses") > 0 && req_json.count("api_keys") > 0)) {
+        res->set_400("Either `ip_addresses` or `api_keys` is required.");
+        return false;
+    }
+
+    if(req_json.count("ip_addresses") > 0) {
+        new_rule.entity_type = RateLimitedEntityType::ip;
+        for(const auto& ip: req_json["ip_addresses"]) {
+            new_rule.entity_ids.push_back(ip);
+        }
+    }
+
+    if(req_json.count("api_keys") > 0) {
+        new_rule.entity_type = RateLimitedEntityType::api_key;
+        for(const auto& api_key: req_json["api_keys"]) {
+            new_rule.entity_ids.push_back(api_key);
+        }
+    }
+
+    if(new_rule.action == RateLimitAction::throttle)
+    {
+        if(req_json.count("max_requests_60s") == 0 || req_json.count("max_requests_1h") == 0) {
+            res->set_400("Both `max_requests_60s` and `max_requests_1h` are required for throttling");
+            return false;
+        }
+
+        if(req_json["max_requests_60s"].get<int>() < 1 || req_json["max_requests_1h"].get<int>() < 1) {
+            res->set_400("Both `max_requests_60s` and `max_requests_1h` must be greater than 0");
+            return false;
+        }
+
+        new_rule.max_requests.minute_threshold = req_json["max_requests_60s"];
+        new_rule.max_requests.hour_threshold = req_json["max_requests_1h"];
+
+        if((req_json.count("auto_ban_threshold_num") > 0 && req_json.count("auto_ban_num_days") == 0) || (req_json.count("auto_ban_threshold_num") == 0 && req_json.count("auto_ban_num_days") > 0)) {
+            res->set_400("Both `auto_ban_threshold_num` and `auto_ban_num_days` are required if either is specified");
+            return false;
+        }
+
+        if(req_json.count("auto_ban_threshold_num") > 0 && req_json.count("auto_ban_num_days") > 0) {
+
+            if(req_json["auto_ban_threshold_num"].get<int>() < 1 || req_json["auto_ban_num_days"].get<int>() < 1) {
+                res->set_400("Both `auto_ban_threshold_num` and `auto_ban_num_days` must be greater than 0");
+                return false;
+            }
+
+            new_rule.auto_ban_threshold_num = req_json["auto_ban_threshold_num"];
+            new_rule.auto_ban_num_days = req_json["auto_ban_num_days"];
+        }
+    }
+
+    if(rateLimitManager->edit_rule_by_id(std::stoull(req->params["id"]), new_rule)) {
+        res->set_200("OK");
+        return true;
+    } else {
+        res->set_400("Rule for one or more entities already exist");
+        return false;
+    }
+
+    return true;
+}
+
+bool del_rate_limit(const std::shared_ptr<http_req>& req, const std::shared_ptr<http_res>& res) {
+    RateLimitManager* rateLimitManager = RateLimitManager::getInstance();
+    uint64_t id = std::stoull(req->params["id"]);
+    const auto& ruleOption = rateLimitManager->find_rule_by_id(id);
+
+    if(!ruleOption.ok()) {
+        res->set_404();
```

Comment:
`ruleOption` could in future return non-404 errors as well. Better to do:

```
res->set(ruleOption.code(), ruleOption.error());
```

Check other places where this has been done and fix those as well.